### ISSUE TITLE
Plurality features

### ISSUE THREAD

**Issue by ellietheoracle** [mirukana/mirage#78]
For users who use mirage as a way to talk from multiple accounts, i.e. systems, etc, it's useful to know who's fronting more, manage switches, etc . ( much like https://pluralkit.me/ ) and since mirage is the only client with multiple account support, as well as allowing tags to send messages, i see it as the quite well suited for this .

some features that would be very nice to see implemented:

- [ ] logging of when each user sends messags
- [ ] graphs to display that data
- [ ] register switches, and have the current open account switch with either a button or a /command or something, e.g. /switch <user> 
- [ ] a "latch" mode, where after a message is sent with a tag, all messages sent after that point would be sent from that account

this is all i can really think of for now, but if i think of anything else i'll add it

**Comment by mirukana** [comment]
>  logging of when each user sends messags

We can have something like a `stats.json` that would contain data like
```json
{
    "@user1:example.com": {
        "sent_messages_count": {
            "2020-06-29": 17
            "2020-06-30": 23
        }
    }
}
```
Is that what you want?

>  graphs to display that data

I wouldn't invest time in this personally, especially when we can't use the [Qt charts module](https://doc.qt.io/qt-5/qtcharts-qmlmodule.html) because it's GPL.
You can work with a JSON file, convert it to other formats and generate graphs with external tools that would do a better job than a chat client.

>  register switches, and have the current open account switch with either a button or a /command or something, e.g. /switch 
> a "latch" mode, where after a message is sent with a tag, all messages sent after that point would be sent from that account

In general, I want to avoid adding commands for things that can be better implemented, more discoverable and faster to use with a keyboard-friendly UI.
Idea to solve these two problems: clicking on your avatar in the composer, or pressing a keybind would open a context menu with a list of your accounts in the room. A filter field would be focused by default. Clicking on one of them or selecting it with keyboard would switch to that account to send messages as (basically would do the same as clicking a room in the left pane's room list).

> ability to show rooms that all accounts are present in only once, with an "all" tag or something at the top of the room list, or just some other way or merging them .

I'm not sure what this means, do you want a way to send one message with multiple accounts at the same time?

**Comment by mirukana** [comment]
Renaming this so that more people can understand the title or find the issue

**Comment by ellietheoracle** [comment]
> We can have something like a `stats.json` that would contain data like
> 
> ```json
> {
>     "@user1:example.com": {
>         "sent_messages_count": {
>             "2020-06-29": 17,
>             "2020-06-30": 23
>         }
>     }
> }

that would be good, yeah .

> I wouldn't invest time in this personally, especially when we can't use the [Qt charts module](https://doc.qt.io/qt-5/qtcharts-qmlmodule.html) because it's GPL.

ok, i might look into this then

> In general, I want to avoid adding commands for things that can be better implemented, more discoverable and faster to use with a keyboard-friendly UI.

if there was a command autocomplete function, like the one suggested for #4 i don't really think it would be that much of an issue 
> Idea to solve these two problems: clicking on your avatar in the composer, or pressing a keybind would open a context menu with a list of your accounts in the room. A filter field would be focused by default. Clicking on one of them or selecting it with keyboard would switch to that account to send messages as (basically would do the same as clicking a room in the left pane's room list).

there's already a list of accounts at the top of the roomlist, clicking on one of those would be fine, i don't really think there's a need for a more complex UI
 
> I'm not sure what this means, do you want a way to send one message with multiple accounts at the same time?

no, i want rooms with the same id to only be listed once, so if skylar and ellie are both in a room, it doesn't need to be listed in both of their accounts, if there was an "all" tag at the top of the roomlist, that would be ideal .

**Comment by mirukana** [comment]
> if there was a command autocomplete function, like the one suggested for #4 i don't really think it would be that much of an issue

There should be one obvious way to do things

> there's already a list of accounts at the top of the roomlist, clicking on one of those would be fine, i don't really think there's a need for a more complex UI

It has a different purpose: a list of *all* your accounts where clicking on one makes the room list scroll to the beginning of it and focus the first room or account settings if collapsed.
The menu would allow quick access to your accounts present in the room (you might have 20 accounts but only 2 of them in X room), and without having to go back to the room list if you're in a small window/mobile mode

> no, i want rooms with the same id to only be listed once, so if skylar and ellie are both in a room, it doesn't need to be listed in both of their accounts, if there was an "all" tag at the top of the roomlist, that would be ideal .

This would further complicate the room list, part of the reasons are explained [here](https://github.com/mirukana/mirage/issues/6#issuecomment-634221918). And under which account would a room be opened, if you clicked one in the "All" section?

**Comment by ellietheoracle** [comment]
> There should be one obvious way to do things

 yeah, fair enough, best leave commands to text functions anyways to keep things simple

> It has a different purpose: a list of _all_ your accounts where clicking on one makes the room list scroll to the beginning of it and focus the first room or account settings if collapsed.
> The menu would allow quick access to your accounts present in the room (you might have 20 accounts but only 2 of them in X room), and without having to go back to the room list if you're in a small window/mobile mode

yeah, that's fair i just think the implementation is important, clicking on the user icon for example would also be a really convenient way to allow users to set their status #73 maybe clicking on the user icon could bring up a quick menu, to change status, change current user, or even open the user's profile page for example .

> This would further complicate the room list, part of the reasons are explained [here](https://github.com/mirukana/mirage/issues/6#issuecomment-634221918). And under which account would a room be opened, if you clicked one in the "All" section?

a sync accounts or tags feature might make that simpler, but whatever the current active user is, would be the one that would be opened by default, and after a switch happens, the new current active user would be the default

**Comment by mirukana** [comment]
> Maybe clicking on the user icon could bring up a quick menu, to change status, change current user, or even open the user's profile page for example

That's a really good idea, it could open a control that combines [`AccountContextMenu`](https://github.com/mirukana/mirage/blob/master/src/gui/MainPane/AccountContextMenu.qml) with the current user switcher.

 > a sync accounts or tags feature might make that simpler, but whatever the current active user is, would be the one that would be opened by default, and after a switch happens, the new current active user would be the default

I would really prefer to wait for Qt to have a proper tree view component before doing this, because it raises too many complicated UI questions. Another one is, once we have room tag sections (favorites, low priority, etc), what happens in the "All" if a room is in multiple accounts but is tagged differently within each of those accounts?

**Comment by ellietheoracle** [comment]
> I would really prefer to wait for Qt to have a proper tree view component before doing this, because it raises too many complicated UI questions. Another one is, once we have room tag sections (favorites, low priority, etc), what happens in the "All" if a room is in multiple accounts but is tagged differently within each of those accounts?

well if the tags in the accounts weren't synced, then the tags of the active account would be applied, but the ability to sync tags would make things less complicated for the user .
